Automatic memory pool eviction. - Kestrel and HTTP.sys (#35822)
dotnet/AspNetCore.Docs: aspnetcore/fundamentals/servers/includes/memory-eviction2.md

## Automatic eviction from memory pool

The memory pools used by Kestrel, IIS, and HTTP.sys automatically evict memory blocks when the application is idle or under low load. The feature runs automatically and doesn't need to be enabled or configured manually.

In versions of .NET earlier than 10, memory allocated by the pool remains reserved, even when not in use. This automatic eviction feature reduces overall memory usage and helps applications stay responsive under varying workloads.

### Use memory pool metrics

The default memory pool used by the ASP.NET Core server implementations includes metrics, which can be used to monitor and analyze memory usage patterns. The metrics are under the name `"Microsoft.AspNetCore.MemoryPool"`.

For information about metrics and how to use them, see <xref:log-mon/metrics/metrics>.

## Manage memory pools

> Besides using memory pools efficiently by evicting unneeded memory blocks, ASP.NET Core provides a built-in [IMemoryPoolFactory](https://source.dot.net/#Microsoft.AspNetCore.Connections.Abstractions/IMemoryPoolFactory.cs) and an implementation for it. It makes the implementation available to your application through dependency injection.

Reviewer: ```suggestion
Besides using memory pools efficiently by evicting unneeded memory blocks, ASP.NET Core provides a built-in [IMemoryPoolFactory](https://source.dot.net/#Microsoft.AspNetCore.Connections.Abstractions/IMemoryPoolFactory.cs) and an implementation for it. ASP.NET Core makes the implementation available to your application through dependency injection.
```
Minor clarification if you choose to add.  Not required.

Author: good catch thanks!

The following code example shows a simple background service that uses the built-in memory pool factory implementation to create memory pools. These pools benefit from the automatic eviction feature:

:::code language="csharp" source="~/fundamentals/servers/snippets/10.x/my-background-service.cs":::

> To use your own memory pool factory, make a class that implements `IMemoryPoolFactory` and register it with dependency injection, as the following example does. Memory pools created this way also benefit from the automatic eviction feature:
> 
> :::code language="csharp" source="~/fundamentals/servers/snippets/10.x/memory-pool-factory.cs":::

Author: ```suggestion
To use a custom memory pool factory, make a class that implements `IMemoryPoolFactory` and register it with dependency injection, as the following example does. Memory pools created this way also benefit from the automatic eviction feature:

:::code language="csharp" source="~/fundamentals/servers/snippets/10.x/memory-pool-factory.cs":::
```

When you're using a memory pool, check the pool's [MaxBufferSize](https://github.com/dotnet/api/system.buffers.memorypool-1.maxbuffersize?view=net-9.0#system-buffers-memorypool-1-maxbuffersize) property.

## Automatic eviction from memory pool

The memory pools used by Kestrel, IIS, and HTTP.sys automatically evict memory blocks when the application is idle or under less load. The feature runs automatically and doesn't need to be enabled or configured manually.

> In versions of .NET earlier than 10 earlier than 10, memory allocated by the pool would remain reserved, even when not in use. This feature releases memory back to the system when the app is idle for a period of time. This eviction reduces overall memory usage and helps applications stay responsive under varying workloads.

Reviewer 2: ```suggestion
In versions of .NET earlier than 10, memory allocated by the pool would remain reserved, even when not in use. This feature releases memory back to the system when the app is idle for a period of time. This eviction reduces overall memory usage and helps applications stay responsive under varying workloads.
```
